### PR TITLE
feat(packaging): bundle xgrammar in the venvstacks export (no-torch stub)

### PR DESCRIPTION
**Pull request by cfbraun**
## Summary

Enables structured output (xgrammar) in the Swift macOS bundle. After #1355 / #1371 the legacy DMG path retired and `packaging/build.py` now only produces the venvstacks Python layers consumed by `apps/omlx-mac/Scripts/build.sh`. Bundle users who try `/v1/chat/completions` with `response_format` currently fall through to "structured output unavailable" — the alternative (`brew install` with `--with-grammar`) pulls real torch (~500 MB) and works, but the bundled-app flow has no equivalent today.

The blocker: xgrammar declares `torch>=1.10.0` as a runtime dep, but oMLX only exercises its non-torch paths (numpy bitmasks + MLX kernel). Module-level `import torch` in `xgrammar.matcher` / `xgrammar.testing` / `xgrammar.contrib.hf` / `tvm_ffi.core` aborts the omlx grammar codepath import on a torch-less install.

## What's in this PR

1. **`omlx/_torch_stub.py`** — a minimal torch surface for xgrammar's import-time references. When real torch is on the path, `install()` finds it via `importlib.util.find_spec("torch")` and bows out as a no-op. Stubbed helpers that would actually call into torch raise a clearly-tagged `RuntimeError "torch.X not supported in omlx stub"` so a future code path that *does* reach into torch-land surfaces with an identifiable error rather than silently misbehaving.

2. **`_install_xgrammar()` in `packaging/build.py`** — downloads `xgrammar` + `apache-tvm-ffi` wheels `--no-deps` (skip torch), extracts them into the `framework-mlx-framework` site-packages layer of the venvstacks export. Hooked into `build_venvstacks()` right after `_install_paroquant()`. Idempotent via an atomic sentinel file with an integrity check (handles SIGKILL / ENOSPC mid-extract). Version constants imported from the stub so the two stay in lockstep.

3. **Post-strip safety check in `_strip_unused_packages()`** — fails the build if any torch artifact survived the strip step. A partial torch (some files but not enough for xgrammar) would make `find_spec("torch")` return truthy, the stub would short-circuit, and xgrammar would import a broken half-torch and fail at runtime. Fail fast at build time.

4. **`install()` call sites** — one each at the xgrammar entry points: `api/grammar.py`, `server.py`, `admin/routes.py`, `tests/conftest.py`. With real torch present each is a no-op.

> Apache-2.0 NOTICE: the previous revision wrote a `NOTICE-third-party.txt` into the old `.app` bundle's Resources dir. That bundle path no longer exists in `packaging/build.py`; Apache-2.0 §4 compliance for the redistributed wheels now belongs in `apps/omlx-mac/` (Swift bundle Resources). Out of scope for this PR.

## Why bundled-app users would want this — and what's NOT affected

| Concern | Reality |
|---|---|
| Bundle size | xgrammar + tvm-ffi ≈ 20–50 MB. Compare to torch (~500 MB) — *net smaller* than the alternative people are avoiding. |
| Stub doesn't cover a torch path oMLX reaches | Audited at write time: bitmasks via numpy + MLX kernel, no torch-backed path is reached. If a future feature does reach one, identifiable `RuntimeError` — not silent. |
| Version drift when xgrammar / tvm-ffi update | `_TARGET_*_VERSIONS` lives in the stub; `_install_xgrammar` imports those constants. They can't drift. `warn_if_unexpected_versions()` logs at startup if reality doesn't match. |
| Real-torch users see new behavior | `install()` early-returns when `find_spec("torch") is not None`. Homebrew `--with-grammar` flow is unaffected. |
| pip users | `install()` is a no-op without real torch on path, but it also doesn't auto-install the stub on the pip-install side — only opts in inside oMLX's own xgrammar entry points. pip users with `pip install omlx[grammar]` failing for lack of torch get exactly the same behavior as today (no change). |

## Test plan

- [x] `pytest tests/test_torch_stub.py` — 17 passed
  - `install()` is a no-op when real torch present
  - `install()` repopulates `sys.modules` correctly when stub is the active torch
  - xgrammar imports successfully against the stub
  - stubbed helpers raise identifiable `RuntimeError` on call
  - thread-safety: concurrent `install()` calls don't double-install
  - version-drift warning when xgrammar / tvm-ffi at runtime don't match `_TARGET_*_VERSIONS`
- [x] Manually verified xgrammar's structured-tag path runs end-to-end against the stub on a torch-less venvstacks-only export (Qwen3.5 / Gemma4 reasoning parsers compile and apply masks).

## Maintenance

Each xgrammar / tvm-ffi version bump = ~30 min:
1. Bump `_TARGET_XGRAMMAR_VERSIONS` / `_TARGET_TVM_FFI_VERSIONS` in `omlx/_torch_stub.py`.
2. Run `pytest tests/test_torch_stub.py` — the stub uses real xgrammar to import-test the stubbed surface.
3. If a new module-level torch reference appeared, add it to `_make_top_level_torch_getattr` or `_build_modules`.

If xgrammar eventually drops torch as a runtime dep (likely), this PR's machinery can be removed in a single follow-up.

### CODE REVIEW

**Comment by fry69** [comment]
I am trying this PR out. It builds and runs without problems (as the new native macOS app!), and `xgrammar` seems to get initialized correctly. There are some warnings in the log though:

```
2026-05-27 12:54:25,781 - omlx.engine_pool - INFO - [-] - Loaded model: Huihui-Qwen3.6-27B-abliterated-mlx-nvfp4 (actual: 15.34GB, estimated: 15.70GB, total: 15.70GB)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.uint16 (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.uint32 (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.uint64 (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float8_e4m3fn (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float8_e4m3fnuz (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float8_e5m2 (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float8_e5m2fnuz (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float8_e8m0fnu (install real torch if this is load-bearing)
2026-05-27 12:54:25,791 - omlx._torch_stub - WARNING - [-] - oMLX torch stub missing attribute: torch.float4_e2m1fn_x2 (install real torch if this is load-bearing)
2026-05-27 12:54:26,074 - omlx.engine.vlm - INFO - [-] - GrammarCompiler initialized for /Users/fry/.omlx/models/Huihui-Qwen3.6-27B-abliterated-mlx-nvfp4
```

They seem harmless, maybe these warnings should only be of `DEBUG` level?

**Comment by cfbraun** [comment]
Thanks for trying it out and the precise log dump — really helpful.

Fixed in `9087d28`. The 9 names you saw are all dtype probes (`tvm_ffi.dtypes` enumerates torch dtypes via `getattr(torch, name)` and catches `AttributeError` — feature detection that falls back gracefully). The WARNING was wrong for those: there's no operator action to take.

Approach: allowlist of known-probed names → DEBUG. Anything outside the allowlist still logs WARNING so a future xgrammar / tvm_ffi reaching for a torch surface we didn't anticipate stays visible in operator logs (which was the original diagnostic intent).

Allowlist seeded with the exact set you observed (`uint16`/`uint32`/`uint64`, `float8_e4m3fn`/`float8_e4m3fnuz`/`float8_e5m2`/`float8_e5m2fnuz`/`float8_e8m0fnu`, `float4_e2m1fn_x2`). New test (`test_known_probe_names_log_at_debug_not_warning`) pins the level distinction so a refactor can't silently re-noise the log.

Could you re-test against the new tip and confirm the log is clean now?

**Comment by fry69** [comment]
Yep, rebuilds fine again, no warnings anymore:

```
2026-05-27 13:27:00,735 - omlx.engine_core - INFO - [-] - Engine started
2026-05-27 13:27:00,775 - omlx.engine.vlm - INFO - [-] - VLM tool calling enabled: parser=qwen3_coder
2026-05-27 13:27:00,777 - omlx.engine.vlm - INFO - [-] - VLMBatchedEngine loaded: /Users/fry/.omlx/models/Huihui-Qwen3.6-27B-abliterated-mlx-nvfp4
2026-05-27 13:27:00,794 - omlx.engine_pool - INFO - [-] - Loaded model: Huihui-Qwen3.6-27B-abliterated-mlx-nvfp4 (actual: 15.35GB, estimated: 15.70GB, total: 15.70GB)
```

**Comment by jundot** [comment]
Thanks for this, and for the unusually thorough write-up. The stub surface, the thread-safe idempotent install, and the post-strip torch guard all look solid to me, and fry69's build confirms structured output comes up clean on the stub-only bundle. The dtype-probe log noise is sorted now too. One item I'll track before the next bundled release: the Apache-2.0 NOTICE for the redistributed xgrammar + tvm-ffi wheels still needs to land in the Swift bundle Resources, as you flagged is out of scope here. Merging this, thanks again.

**Comment by fry69** [comment]
FYI: `xgrammer` loaded fine during testing this PR when it was not merged, but building current `main` (04da6ff68edc1d2c0e34c0fbbee4ebeca38cd0fe) and running it produces this warning and structured outputs do not seem to get respected in my test like they do with e.g. `ollama` with the same model.

```
omlx.engine.batched - INFO - [-] - Structured output is not available in the DMG version (xgrammar requires torch which significantly increases app size). Use the pip or Homebrew version for structured output support.
```

Command to build:
```shell
time apps/omlx-mac/Scripts/build.sh release --rebuild-donor
```

I cross checked with the `pip` install and a fresh `.venv` environment and it also shows that `xgrammar` does not get loaded/detected:

```
omlx.engine.batched - INFO - [-] - Structured output requires xgrammar. Install with: pip install 'omlx[grammar]'
```

Am I missing something? Should this not work now out of the box?
